default-placeholders/index.md

```diff
@@ -0,0 +1,269 @@
+   we propose making the handler optional.
+   This is the same constructor currently allowed for host buffers.
+
+```cpp
+template <typename dataT,
+          int dimensions,
+          access::mode accessMode,
+          access::target accessTarget = access::target::global_buffer,
+          access::placeholder isPlaceholder = access::placeholder::false_t>
+class accessor {
+ public:
+  ...
+
+  // 1
+  // Only available when ((accessTarget == access::target::global_buffer) ||
+  //                      (accessTarget == access::target::constant_buffer))
+  accessor() noexcept;
+  
+  // 2
+  // Only available when: ((accessTarget == access::target::global_buffer) ||
+  //                       (accessTarget == access::target::constant_buffer) ||
+  //                       (accessTarget == access::target::host_buffer)) &&
+  //                      (dimensions == 0)
+  accessor(buffer<dataT, 1> &bufferRef);
+
+  // 3
+  // Only available when: ((accessTarget == access::target::global_buffer) ||
+  //                       (accessTarget == access::target::constant_buffer) ||
+  //                       (accessTarget == access::target::host_buffer)) &&
+  //                      (dimensions > 0)
+  accessor(buffer<dataT, dimensions> &bufferRef,
+           range<dimensions> accessRange,
+           id<dimensions> accessOffset = {});
+};
+```
+
+### New `accessor` member functions
+
+In order to query the accessor for its status,
+we propose new member functions to the `accessor class`:
+
+1. `is_null` - returns `true` if the accessor has been default constructed,
+   which is only possible with placeholders.
+   Not having an associated buffer is analogous to a null pointer.
+1. `has_handler` - returns `true` if the accessor is associated
+   with a command group `handler`.
+   Will only be `false` with host accessors and placeholder accessors.
+   This replaces the `is_placeholder` member function.
+
+```cpp
+template <typename dataT,
+          int dimensions,
+          access::mode accessMode,
+          access::target accessTarget = access::target::global_buffer,
+          access::placeholder isPlaceholder = access::placeholder::false_t>
+class accessor {
+ public:
+  ...
+
+  // 1
+  bool is_null() const noexcept;
+  
+  // 2
+  bool has_handler() const noexcept;
+};
+```
+
+## Examples
+
+Simple vector addition:
+
+```cpp
+std::vector<int> a{1, 2, 3, 4, 5};
+std::vector<int> b{6, 7, 8, 9, 10};
+
+using read_acc =
+  accessor<int, 1, access::mode::read, access::target::global_buffer>;
+using write_acc =
+  accessor<int, 1, access::mode::discard_write, access::target::global_buffer>;
+
+read_acc accA;
+read_acc accB;
+write_acc accC;
+
+// Sanity checks
+assert(accA.is_null());
+assert(!accA.has_handler());
+
+const auto N = a.size();
+const auto bufRange = range<1>(N);
+
+queue myQueue;
+
+// Create a buffer and copy `a` into it
+buffer<int> bufA{bufRange};
+
+accA = read_acc{bufA};
```

Comment:
what's a practial sense of an accessor creation w/o buffer, if we re-create the same accessor later?
M.b. it would make sense if we can "bind" a buffer with an "empty" accessor?

Reply:
1. The main use case is for the accessor to be default constructible, so you could easily reuse the same data structures with different buffers. By assigning to it later you are effectively binding it. We could also have an `accessor::bind` member function, which would achieve the same thing as `operator=`. Or are you suggesting a `buffer::bind` member function?
1. If I understand this correctly, you would capture a fully bound accessor once, store it on the host, and use the same accessor in multiple command groups and kernel invocations? I believe that's a separate, and larger, problem - if this proves to be a very useful use case we can look more into it. It seems a lot like USM, though. Also without USM, I don't think pointers are guaranteed to be persistent between kernel invocations, at least not in OpenCL.

Reply:
1. Having an accessor::bind(buffer) member function is OK.
2. I mean a situation when we have bound accessor with SYCL buffer once, we cannot get SYCL buffer back... What should I do for getting result on the host in the following case?

```
template<typename AccTypeA, typename AccTypeB>
void do_something_and_print_result(AccTypeA accA, AccTypeA accA)
{
   ....
    myQueue.submit([&](handler &cgh) {
       cgh.require(accA);
       cgh.require(accB);
       cgh.copy(accA, accB);
   });
   ...
   // For example, here we want a result access
   // But, in that scope we have not cl::sycl::buffer object (to get a host accessor), and there is 
   //not any way to get it.  Would be nice if there is some way to re-bind placholder accessor: 

/*   accB.rebind_to_host();  //here accB "becomes"  a host accessor  */

   for(auto i = 0; i < accB.get_count(); ++i)
       std::cout << accB[i] << " ";
   std::cout << std::endl;
   ...
}
```

Reply:
Sorry, I'm not able to access that issue.

That definitely sounds like a compelling use case. Modifying an accessor to change it from a device to a host one can be a bit problematic, especially since they're very different types. We briefly discussed that we could allow constructing a host accessor from a placeholder accessor. So your example would become - with this proposal and some of https://github.com/codeplaysoftware/standards-proposals/pull/100 - like this:

```cpp
template<typename AccTypeA, typename AccTypeB>
void do_something_and_print_result(AccTypeA accA, AccTypeB accB)
{
   ....
    myQueue.submit([&](handler &cgh) {
       cgh.require(accA);
       cgh.require(accB);
       cgh.copy(accA, accB);
   });
   ...
   { // Remember to scope host accessors because they're essentially locks
       host_accessor hostAcc{accB};
       // or: auto hostAcc = accB.get_host_access();
       for(auto i = 0; i < hostAcc.get_count(); ++i)
           std::cout << hostAcc[i] << " ";
       std::cout << std::endl;
       ...
   }
}
```

We do not have clear consensus on whether this should be permitted or not, but initial investigation reveals this might be OK.

Reply:
1. Sounds great! "Constructing a host accessor from a placeholder accessor" should be enough for our use cases...  Thank you.
Does some formal step require from me to it will be adopted in the spec? Or that example is enough?
2. I'll try to clarify how you can get an access to https://gitlab.devtools.intel.com/SYCL/specbugs/issues/65

Reply:
I've added your example to the proposal, if that's OK, and added a host accessor constructor from a placeholder, and also a `get_host_access` function to placeholders.

For spec adoption, proposals are usually brought to the Khronos group, either on GitHub or internally, and your support at that point helps a lot.

Reply:
Good! Thank you.

Reply:
I faced  to another issue connected with usage of a placeholder accessor - there is not a way(API) to change access mode for given accessor. Please have a look following example:
template<typename AccTypeA, typename AccTypeB>
void do_something_and_print_result(AccTypeA accA, AccTypeB accB)
{
   ....
    myQueue.submit([&](handler &cgh) {
       cgh.require(accA);
       cgh.require(accB);
       cgh.copy(accA, accB);
   });
   ...
   // For example, here we want to do some computation  for data "B", a kind of reduction, for example. We need just 
   //"Read" access for "B", but passed accB has "Write" access mode. So, would be great to have possibility to create (or change) a new accessor with necessary access mode, like this:

read_accessor accB_read{accB}; 

   ...
}

Reply:
https://github.com/codeplaysoftware/standards-proposals/pull/100 talks about allowing accessors to be implicitly converted to `accessor<const dataT>` and also to a read-only accessor (because read-only should be semantically the same as `const dataT`).

I'm not sure how general this should be, because even just going in the other direction gets very tricky very fast, you get into problems with multiple accessor resolution.

Reply:
Thanks! It seems that enough....

More one important use-case which I faced to:

```cpp
template<typename T /* this is some accessor */ >
void foo(T& in)
{
   // For example, here we want to do some computation  for part of data - we want to skip first N elements. So, we 
   // would like to set "offset = N", but there is no API for that(!)
   
   in.set_offeset(N); 
   //or accessor new_acc(in, N);

   myQueue.submit([&](handler &cgh) {
     ...
   }
   ...

}
```